cirq-google/cirq_google/circuit_serializer.py

```diff
@@ -0,0 +1,657 @@
+                will be checked for a given type in the order specified here.
+                This allows for a given gate type to be serialized into
+                different serialized form depending on the parameters of the
+                gate.
+            deserializers: The OpDeserializers to convert serialized
+                forms of gates or circuits into Operations.
+        """
+        self.gate_set_name = gate_set_name
+
+    def serialize(
+        self,
+        program: cirq.Circuit,
+        msg: Optional[v2.program_pb2.Program] = None,
+        *,
+        arg_function_language: Optional[str] = None,
+        use_constants: bool = True,
+    ) -> v2.program_pb2.Program:
+        """Serialize a Circuit to cirq_google.api.v2.Program proto.
+
+        Args:
+            program: The Circuit to serialize.
+            msg: An optional proto object to populate with the serialization
+                results.
+            arg_function_language: The `arg_function_language` field from
+                `Program.Language`.
+            use_constants: Whether to use constants in serialization. This is
+                required to be True for serializing CircuitOperations.
+        """
+        if msg is None:
+            msg = v2.program_pb2.Program()
+        msg.language.gate_set = self.gate_set_name
+        if isinstance(program, cirq.Circuit):
+            constants: Optional[List[v2.program_pb2.Constant]] = [] if use_constants else None
+            raw_constants: Optional[Dict[Any, int]] = {} if use_constants else None
+            self._serialize_circuit(
+                program,
+                msg.circuit,
+                arg_function_language=arg_function_language,
+                constants=constants,
+                raw_constants=raw_constants,
+            )
+            if constants is not None:
+                msg.constants.extend(constants)
+            if arg_function_language is None:
+                arg_function_language = arg_func_langs._infer_function_language_from_circuit(
+                    msg.circuit
+                )
+        else:
+            raise NotImplementedError(f'Unrecognized program type: {type(program)}')
+        msg.language.arg_function_language = arg_function_language
+        return msg
+
+    def serialize_op(
+        self,
+        op: cirq.Operation,
+        msg: Union[None, v2.program_pb2.Operation, v2.program_pb2.CircuitOperation] = None,
+        **kwargs,
+    ) -> Union[v2.program_pb2.Operation, v2.program_pb2.CircuitOperation]:
+        """Disambiguation for operation serialization."""
+        if msg is None:
+            if op.gate is not None:
+                return self.serialize_gate_op(op, msg, **kwargs)
+            if hasattr(op.untagged, 'circuit'):
+                return self.serialize_circuit_op(op, msg, **kwargs)
+            raise ValueError(f'Operation is of an unrecognized type: {op!r}')
+
+        if isinstance(msg, v2.program_pb2.Operation):
```

Comment:
It feels a bit awkward to have to check at runtime to match types between the proto message and the operation itself. I'd suggest having `_serialize_circuit` call `serialize_gate_op` or `serialize_circuit_op` directly since it already checks whether or not a given op is a `CircuitOperation`. I suppose this method could stay if you want it to be available to call directly (e.g. for tests), but I think it'd probably be fine to remove.

Also, I'd suggest moving `_serialize_circuit` up to be immediately after the `serialize` method; it's a bit hard to follow the flow when that internal method is located down after all the deserialization code.

Reply:
Done.